effect/shared/src/main/scala/com/monovore/decline/effect/CommandIOApp.scala

```diff
@@ -1,21 +1,26 @@
 package com.monovore.decline.effect
 
-import cats.effect.{IO, IOApp, ExitCode}
+import cats.effect.implicits._
+import cats.effect.{Effect, ExitCode, IO, IOApp}
 import cats.implicits._
-
 import com.monovore.decline._
 
-abstract class CommandIOApp(
+import scala.language.higherKinds
+
+abstract class CommandIOApp[F[_]: Effect](
     name: String,
     header: String,
     helpFlag: Boolean = true,
     version: String = ""
 ) extends IOApp {
 
-  def main: Opts[IO[ExitCode]]
+  def main: Opts[F[ExitCode]]
 
   private[this] def command: Command[IO[ExitCode]] = {
-    val showVersion = {
+    val mainCommand: Opts[IO[ExitCode]] =
+      main.map(_.toIO)
```

Comment:
So FWIW this isn't exactly what I expected after our discussion on the issue. (Probably my fault!) I had understood it as wanting to use [zio.App](https://zio.dev/docs/overview/overview_running_effects) and its runtime system / etc. with `decline`. I'd been imagining some `def run[F[_] : Effect, A](command: Command[A], args: List[String]): F[A]` method that did the status-setting and error-printing, which could be used a la:

```scala
object MyApp extends zio.App {
  def command: ZIO[ZEnv, Nothing, Int] = ???
  def run(args: List[String]) = CommandIOApp.run(command, args)
}
```

I'm not hung up on that idea, though eg. it's not clear to me if there are downsides for ZIO people using IOApp (does the runtime stuff work out?)

Though in its current state this is mostly saving users a call to `.map(_.toIO)` themselves... which is not nothing but maybe not enough to justify the switch. Thoughts?

Reply:
Well actually I didn't fully understood your idea until now, so definitely not your fault.
About using IOApp with ZIO runtime: when converting an Effect to IO, cats-effect runs the provided effect prematurely via unsafeRunSync, so breaking the idea of running effects only at "the end of the world". That is known to be non-ideal, but other (cats-effect based) major libraries do this at this  time (http4s) and it works to some extent.

About the proposed `def run[F[_]: Effect, A]`: could you please elaborate a bit more? In fact, I am still trying to wrap my head around your example. Are you suggesting to place it in the companion object? If yes, we would lose the showVersion part (due to already requiring Command and not Opts as a parameter) and need to replicate in a similar but polymorphic fashion the printHelp part. I can try to take at stab at it if this matches your idea, just wanted to be sure it is the way you want to go 😄